beer-garden/beer-garden
# #1198 - Added trusted header login handler

Closes #1198

Benefits from, but does not strictly require, beer-garden/brewtils#377.

This PR does the following:
* Adds a `TrustedHeaderLoginHandler` (replaces the stubbed out CertificateLoginHandler) which will use configuration defined headers to determine the requesting user and the role assignments that they should have.
* Adds configuration for specifying which login handlers should be enabled.

## Testing Instructions
* Update the `auth` section of your garden config to look something like:
  ```yaml
  auth:
    authentication_handlers:
      basic:
        enabled: true
      trusted_header:
        enabled: true
        user_groups_header: bg-user-groups
        username_header: bg-username
        create_users: true
    enabled: false
    guest_login_enabled: false
    role_definition_file: /home/brewmeister/beer-garden/conf/roles.yaml
    group_definition_file: /home/brewmeister/beer-garden/conf/groups.yaml
    token_secret: WHATSASECRET
  ```
  * The paths for `role_definition_file` and `group_definition_file` should be updated to reflect the location of those files in your environment.
  * There is a new `example_configs/groups.yaml` that can be used for the `group_definition_file` setting.
* Hit the token endpoint, specifying the headers you have defined in the config:
  ```shell
  curl -X POST -H "bg-username: testuser" -H "bg-user-groups: GLOBAL_SUPERUSER" http://localhost:8080/api/v1/token
  ```
  * You should get tokens back from that request.
* Decode the access token. The user should have the permissions associated with the role mapped to the group supplied in the `bg-user-groups` header (`GLOBAL_SUPERUSER` in the above example).  For an easy way to get and decode the token, you can do:
  ```shell
  export TOKEN=`curl -X POST -H "bg-username: testuser" -H "bg-user-groups: GLOBAL_SUPERUSER" http://localhost:8080/api/v1/token | jq -r .access`
  python -c 'import os, jwt, json; print(json.dumps(jwt.decode(os.getenv("TOKEN"), options={"verify_signature": False})))' | jq .
  ```
